spec/Section 6 -- Execution.md

```diff
@@ -359,17 +348,212 @@ Unsubscribe(responseStream):
 
 ## Executing Selection Sets
 
-To execute a _selection set_, the object value being evaluated and the object
-type need to be known, as well as whether it must be executed serially, or may
-be executed in parallel.
+The process of executing a GraphQL operation is to recursively execute every
+selected field in the operation. To do this, first all initially selected fields
+from the operation's top most _root selection set_ are collected, then each
+executed, then of those all subfields are collected, then each executed. This
+process continues until there are no more subfields to collect and execute.
+
+### Executing the Root Selection Set
+
+:: A _root selection set_ is the top level _selection set_ provided by a GraphQL
+operation. A root selection set always selects from a root type.
 
-First, the selection set is turned into a grouped field set; then, each
-represented field in the grouped field set produces an entry into a result map.
+To execute the root selection set, the initial value being evaluated and the
+root type must be known, as well as whether it must be executed serially, or may
+be executed in parallel (see
+[Normal and Serial Execution](#sec-Normal-and-Serial-Execution).
 
-ExecuteSelectionSet(selectionSet, objectType, objectValue, variableValues):
+Executing the root selection set works similarly for queries (parallel),
+mutations (serial), and subscriptions (where it is executed for each event in
+the underlying Source Stream).
+
+First, the _selection set_ is collected into a _grouped field set_ which is then
+executed, returning the resulting {data} and {errors}.
+
+ExecuteRootSelectionSet(variableValues, initialValue, objectType, selectionSet,
+executionMode):
 
 - Let {groupedFieldSet} be the result of {CollectFields(objectType,
   selectionSet, variableValues)}.
+- Let {data} be the result of running {ExecuteGroupedFieldSet(groupedFieldSet,
+  objectType, initialValue, variableValues)} _serially_ if {executionMode} is
+  {"serial"}, otherwise _normally_).
+- Let {errors} be the list of all _execution error_ raised while executing the
+  selection set.
+- Return an unordered map containing {data} and {errors}.
+
+### Field Collection
+
+Before execution, the _selection set_ is converted to a _grouped field set_ by
+calling {CollectFields()}. This ensures all fields with the same response name,
+including those in referenced fragments, are executed at the same time.
+
+:: A _grouped field set_ is a map where each entry is a _response name_ and a
+list of selected fields that share that _response name_ (the field alias if
+defined, otherwise the field's name).
```

Comment:
If we choose not to rename "grouped field set", we should at least define it in terms of a set:

```suggestion
:: A _grouped field set_ is a map where each entry is a _response name_ and an
ordered set of selected fields that share that _response name_ (the field alias
if defined, otherwise the field's name).
```

Reply:
If we accept this, we should also make a few other changes such as:

```diff
-Initialize groupedFields to an empty ordered map of lists.
+Initialize groupedFields to an empty ordered map of ordered sets.
```

Reply:
There's no functional change between a list or a set in this case since the algorithm never attempts to add the same item to the set/list?

Reply:
Yes indeed, list or ordered set is equivalent here since it will never include duplicates. The main reason to change it to set is because we call it a "grouped field set"; a "field set" being `OrderedSet<Field>`, and grouped being a map grouped by response key: `Map<string, OrderedSet<Field>>`.

Reply:
This might be a simpler solve actually:

```suggestion
:: A _grouped field set_ is a map where each entry is a _response name_ and a
list of selected fields that share that _response name_ (the field alias if
defined, otherwise the field's name).

Note: No entry in the list of selected fields will be duplicated, hence we call
it a "field set" even though in the algorithms we use a list structure for it.
```